keps/sig-instrumentation/1748-pod-resource-metrics/README.md

```diff
@@ -0,0 +1,497 @@
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal and how do we mitigate.  Think broadly.
+For example, consider both security and how this will impact the larger
+kubernetes ecosystem.
+
+How will security be reviewed and by whom?
+
+How will UX be reviewed and by whom?
+
+Consider including folks that also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+### Expose new metrics
+
+The scheduler will expose an optional set of metrics series that capture the resources a pod is considered to be "consuming" from a scheduling perspective at the current time.
+
+The series reported will be consistent with `kube_node_status_allocatable` to simplify queries that compare used capacity and available capacity.
+
+* `kube_pod_resource_requests` contains series for the consumed requested resources for a pod at a given time.
+  * The value is the quantity of the limit in `spec.containers.resources.requests`
+  * It has the following labels:
+    * `pod` - the name of the pod
+    * `namespace` - the namespace of the pod
+    * `resource` - the name of the resource as described in `spec.containers.resources.requests`
+    * `units` - the units of the resource type as inferred from the resource name in `spec.containers.requests`, such as `bytes`, `cores`. Empty if this is a unitless resource.
+* `kube_pod_resource_limits` contains series for the consumed limit resources for a pod at a given time.
+  * The value is the quantity of the limit in `spec.containers.resources.limits`
+  * It has the following labels:
+    * `pod` - the name of the pod
+    * `namespace` - the namespace of the pod
+    * `resource` - the name of the resource as described in `spec.containers.resources.limits`
+    * `units` - the units of the resource type as inferred from the resource name in `spec.containers.limits`, such as `bytes`, `cores`. Empty if this is a unitless resource.
+* `kube_pod_resource_requests_lifecycle` contains series for the consumed requested resources for a pod at a given time when the pod is in the provided lifecycle phase. All pods are in one of the lifecycle phases, but series will only be returned for the current phase.
+  * The value is the quantity of the limit in `spec.containers.resources.requests`
+  * It has the following labels:
+    * `lifecycle` - this is set to the current lifecycle phase as described in the resource model - `Pending`, `Initializing`, `Running`, `Completed`
```

Comment:
Are `Pending` and `Completed` pods considered to consume 0 resources, or a different amount?

Reply:
Oops, realized I left the node label out.  Pending vs Completed is that Pending reserves resources but not from a node, and completed does not reserve resources but is on a node (until deletion)

Reply:
Thanks.  For completed, what would someone use that stream for?  Completed pods don't generally occupy resources on the node, or count against quotas.  To be explicit, I can see why we wouldn't want to include Completed pods in Running, but not why we would want to use the stream itself.

Reply:
>  Completed pods don't generally occupy resources on the node, or count against quotas.

They use disk space for logs today that probably in the future we need to track and manage.  Completed pods count against total pod object count from an etcd defensiveness (and has to for quota to satisfy safety guarantees in the apiserver since quota's job is to block resource exhaustion).

I should clarify that both resource_requests has to zero out requests for completed pods in the doc, i left that detail out and it's significant. Otherwise `resource_requests` is useless for answering questions of "how many resources are accounted to a node".  Completed resource requests shows you the resources of that completed pod, so that you can reason about very short or fast jobs.

Reply:
> They use disk space for logs today that probably in the future we need to track and manage.

True, although it is considered best-effort.  The kubelet can remove logs from dead containers during out-of-resource handling.  But the point about counting towards quotas is well-taken.  TIL